addons/interactions/src/instrument.ts

```diff
@@ -0,0 +1,239 @@
+    args: call.args.map((arg) => {
+      if (iframeState.callRefsByResult.has(arg)) {
+        return iframeState.callRefsByResult.get(arg);
+      }
+      if (arg instanceof global.window.HTMLElement) {
+        const { prefix, localName, id, classList, innerText } = arg;
+        const classNames = Array.from(classList);
+        return { __element__: { prefix, localName, id, classNames, innerText } };
+      }
+      return arg;
+    }),
+  };
+
+  // Mark any ancestor calls as "chained upon" so we won't attempt to defer it later.
+  call.path.forEach((ref: any) => {
+    if (ref?.__callId__) sharedState.chainedCallIds.add(ref.__callId__);
+  });
+
+  try {
+    // An earlier, non-interceptable call might have forwarded an exception.
+    if (iframeState.forwardedException) {
+      throw iframeState.forwardedException;
+    }
+
+    const result = fn(
+      // Wrap any callback functions to provide a way to access their "parent" call.
+      ...call.args.map((arg: any) => {
+        if (typeof arg !== 'function' || Object.keys(arg).length) return arg;
+        return (...args: any) => {
+          const prev = iframeState.parentCallId;
+          iframeState.parentCallId = call.id;
+          const res = arg(...args);
+          iframeState.parentCallId = prev;
+          return res;
+        };
+      })
+    );
+    channel.emit(EVENTS.CALL, { ...info, state: CallStates.DONE });
+
+    // Track the result so we can trace later uses of it back to the originating call.
+    // Primitive results (undefined, null, boolean, string, number, BigInt) are ignored.
+    if (result && ['object', 'function', 'symbol'].includes(typeof result)) {
+      iframeState.callRefsByResult.set(result, { __callId__: call.id, retain: call.retain });
+    }
+
+    return result;
+  } catch (e) {
+    if (e instanceof Error) {
+      const { name, message, stack } = e;
+      const exception = { name, message, stack };
+      channel.emit(EVENTS.CALL, { ...info, state: CallStates.ERROR, exception });
+
+      // Always track errors to their originating call.
+      iframeState.callRefsByResult.set(e, { __callId__: call.id, retain: call.retain });
+
+      // We need to throw to break out of the play function, but we don't want to trigger a redbox
+      // so we throw an ignoredException, which is caught and silently ignored by Storybook.
+      if (call.interceptable) {
+        throw IGNORED_EXCEPTION;
+      }
+
+      // Non-interceptable calls need their exceptions forwarded to the next interceptable call.
+      iframeState.forwardedException = e;
+      return e;
+    }
+    throw e;
+  }
+}
+
+function intercept(fn: Function, call: Call) {
+  // For a "jump to step" action, continue playing until we hit a call by that ID.
+  // For chained calls, we can only return a Promise for the last call in the chain.
+  const isChainedUpon = sharedState.chainedCallIds.has(call.id);
+  if (!sharedState.isDebugging || isChainedUpon || sharedState.playUntil) {
```

Comment:
```suggestion
  // For chained calls, we can only return a Promise for the last call in the chain.
  // For calls that are passed as arguments to later calls, we only stop at the later call.
  const isChainedUpon = sharedState.chainedCallIds.has(call.id);
  const isPassedAsArg = /* ? */
  if (!sharedState.isDebugging || isChainedUpon || isPassedAsArg || sharedState.playUntil) {
```

We discussed how this behaviour is inconsistent to the `fold` function above.

Reply:
The problem was that an interceptable call that's passed as argument to another interceptable call, would return a Promise while debugging, causing the outer call to break if it doesn't expect a Promise. Additionally, if the inner call would throw an exception, this would never appear in the UI because this inner call didn't have its own row to display the error on.

This was resolved by changing the prev/next logic to walk the list of interactions (the visible rows) rather than the raw log. It uses `playUntil` to keep playing until it hits a call that has its own row, ensuring that we never pause at the nested call. I thought I had the exception handling nailed by forwarding exceptions to the next interceptable call, but in case of a nested interceptable call, this isn't going to work, so I'll revisit that.

Reply:
I gave the exception situation some more thought, and I don't think we can do anything about it. What'll happen is that the nested call will get its own row in the table, and that row will show the exception, and there won't be any further rows in the list, because we've broken out of the play function by throwing.
We could catch the exception and not re-throw, so that we can check if there's a "parent" call, but that call might very well break as a result of it receiving an Error rather than whatever it normally expects. Moreover, in between the nested call (that throws) and the parent call, there might be other calls (that may or may not break themselves).
I'm going to consider this problem resolved, since we have no better alternative.